Lib/functools.py

```diff
@@ -192,6 +192,208 @@ def total_ordering(cls):
+
+    def is_active(self):
+        """Return True if more progress can be made and ``False`` otherwise.
+
+        Progress can be made if cycles do not block the resolution and either there
+        are still nodes ready that haven't yet been returned by "get_ready" or the
+        number of nodes marked "done" is less than the number that have been returned
+        by "get_ready".
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+        return self.nfinished < self.npassedout or bool(self.ready_nodes)
+
+
+    def done(self, node):
+        """Marks a nodes returned by "get_ready" as processed.
+
+        This method unblocks any successor of *node* for being returned in the future
+        by a call to "get_ready"
+
+        Raises :exec:`ValueError` if *node* has already been marked as processed by a
+        previous call to this method, if *node* was not added to the graph by using
+        "add" or if called without calling "prepare" previously or if node has not
+        yet been returned by "get_ready".
+        """
+
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        n2i = self.node2info
+
+        # Check if we know about this node (it was added previously using add()
+        if (nodeinfo := n2i.get(node)) is None:
+            raise ValueError(f"node {node!r} was not added using add()")
+
+        # If the node has not being returned (marked as ready) previously, inform the user.
+        stat = nodeinfo.npredecessors
+        if stat != _NODE_OUT:
+            if stat >= 0:
+                raise ValueError(f"node {node!r} was not passed out (still not ready)")
+            elif stat == _NODE_DONE:
+                raise ValueError(f"node {node!r} was already marked done")
+            else:
+                assert False, f"node {node!r}: unknown status {stat}"
+
+        # Mark the node as processed
+        nodeinfo.npredecessors = _NODE_DONE
+
+        # Go to all the successors and reduce the number of predecessors, collecting all the ones
+        # that are ready to be returned in the next get_ready() call.
+        for successor in nodeinfo.successors:
+            successor_info = n2i[successor]
+            successor_info.npredecessors -= 1
+            if successor_info.npredecessors == 0:
+                self.ready_nodes.append(successor)
+        self.nfinished += 1
+
+    def _find_cycle(self):
+        n2i = self.node2info
+        stack = []
+        itstack = []
+        seen = set()
+        node2stacki = {}
+
+        for node in n2i:
+            if node in seen:
+                continue
+
+            while True:
+                # If we have seen already the node and is in the
+                # current stack we have found a cycle.
+                if node in seen and node in node2stacki:
```

Comment:
This changed the logic of the code you started from, but not in a good way.   If `node in seen` but it's not in `node2stacki`, we do _not_ want to go on to redundantly add `node` to `seen` (and so on, and so on) again.  The original nested logic was correct:

```python
if node in seen:
    if node in node2stacki:
        return ...
    # else go on to get next successor
else:
    seen.add(node)
    ...
```

Reply:
Whoops! Apologies and thanks for the catch

Reply:
Done in 301a459. Is 4 AM here....I should go to bed :sweat_smile: